src/main/java/be/seeseemelk/mockbukkit/block/state/BlockStateMock.java

```diff
@@ -293,4 +294,22 @@ public static BlockStateMock mockState(@NotNull Block block)
 			return new BlockStateMock(block);
 		}
 	}
+
+	@Override
+	public boolean equals(Object other) {
+		if(other == null) {return false;}
+		if(!(other instanceof BlockState)) {return false;}
+
+		BlockState otherState = (BlockState) other;
+		if(otherState.getType() != getType()) {return false;}
+		if(otherState.isPlaced() != isPlaced()) {return false;}
+		if(!isPlaced()) {return true;}
+
+		return otherState.getBlock().getLocation().equals(getBlock().getLocation());
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(material, block == null ? null : block.getLocation());
+	}
```

Comment:
1. This method breaks the contract here as .hashCode() should always be inline with what .equals() does.
2. Look at the Spigot implementation, it not only takes the location (world + position) into account, but also the `BlockData`, so when mocking this behaviour should be mirrored.

```java
    @Override
    public int hashCode() {
        int hash = 7;
        hash = 73 * hash + ((this.world != null) ? this.world.hashCode() : 0);
        hash = 73 * hash + ((this.position != null) ? this.position.hashCode() : 0);
        hash = 73 * hash + ((this.data != null) ? this.data.hashCode() : 0);
        return hash;
    }
```

Reply:
BlockPosition isn't a class that I can use - it's NMS. So using the location of the internal Block makes more sense. (or null if none)
BlockData isn't yet implemented, so I can't use that.

I can change it to remove the Objects.hash call and try to mirror the Bukkit method as closely a possible in that way if you wish, but other than that I can't get much closer without implementing BlockData.

Reply:
There is a `BlockDataMock`, however it is currently not being used by the `BlockStateMock`. Some work would be required to instantiate the correct `BlockDataMock` for any given `BlockStateMock`.

Reply:
Ah there is, I see. I just assumed that there wasn't, since I saw that trying to get it from a state mock just threw a not implemented exception.
I don't have much time atm, rewriting other stuff, but maybe I suppose.
Thanks for the insight